esmvalcore/preprocessor/_trend.py

```diff
@@ -0,0 +1,185 @@
+def _get_slope(y_arr, x_arr):
+    """Calculate slope between X and Y array."""
+    (x_arr, y_arr) = _mask_xy(x_arr, y_arr)
+
+    # If less than 2 points, slope calculation not possible
+    if len(y_arr) < 2:
+        return np.nan
+
+    # Calculate slope
+    return _slope(x_arr, y_arr)
+
+
+def _get_slope_stderr(y_arr, x_arr):
+    """Calculate standard error of slope between X and Y array."""
+    (x_arr, y_arr) = _mask_xy(x_arr, y_arr)
+
+    # If less than 2 points, slope stderr calculation not possible
+    if len(y_arr) < 2:
+        return np.nan
+
+    # Calculate standard error of slope
+    if len(y_arr) == 2:
+        return 0.0
+    dof = len(y_arr) - 2
+    x_mean = x_arr.mean()
+    y_mean = y_arr.mean()
+    slope = _slope(x_arr, y_arr)
+    intercept = y_mean - slope * x_mean
+    y_estim = slope * x_arr + intercept
+    slope_stderr = np.sqrt(((y_arr - y_estim)**2).sum() / dof /
+                           ((x_arr - x_mean)**2).sum())
+    return slope_stderr
+
+
+def _set_trend_units(cube, coord):
+    """Set correct trend units for cube."""
+    coord_units = coord.units
+    if coord_units.is_time_reference():
+        coord_units = Unit(coord_units.symbol.split()[0])
+    invalid_units = any([cube.units is None, cube.units.is_unknown(),
+                         cube.units.is_no_unit(), coord_units.is_no_unit()])
+    if not invalid_units:
+        cube.units /= coord_units
+
+
+def linear_trend(cube, coordinate='time'):
+    """Calculate linear trend of data along a given coordinate.
+
+    The linear trend is defined as the slope of an ordinary linear regression.
+
+    Parameters
+    ----------
+    cube : iris.cube.Cube
+        Input data.
+    coordinate : str, optional (default: 'time')
+        Dimensional coordinate over which the trend is calculated.
+
+    Returns
+    -------
+    iris.cube.Cube
+        Trends.
+
+    Raises
+    ------
+    iris.exceptions.CoordinateNotFoundError
+        The dimensional coordinate with the name ``coordinate`` is not found in
+        ``cube``.
+
+    """
+    coord = cube.coord(coordinate, dim_coords=True)
+
+    # Construct aggregator and calculate trend
+    def call_func(data, axis, x_data):
+        """Calculate trend."""
+        trend_arr = np.apply_along_axis(_get_slope, axis, data, x_data)
+        trend_arr = np.ma.masked_invalid(trend_arr)
+        return trend_arr
+
+    def lazy_func(data, axis, x_data):
+        """Calculate trend lazily."""
+        trend_arr = da.apply_along_axis(
+            _get_slope, axis, data, x_data, dtype=data.dtype, shape=())
+        trend_arr = da.ma.masked_invalid(trend_arr)
+        return trend_arr
+
+    aggregator = iris.analysis.Aggregator('trend', call_func,
+                                          lazy_func=lazy_func,
+                                          x_data=coord.points)
+    cube = cube.collapsed(coord, aggregator)
+
+    # Adapt metadata
+    if cube.var_name is not None:
+        cube.var_name += '_trend'
```

Comment:
I'm not sure if an `_` is a valid character for a `short_name`, because that would break splitting the filenames on `_` to get the facets of the DRS.